src/domain/mongoose_domain_api.erl

```diff
@@ -0,0 +1,130 @@
+                false ->
+                    {error, service_disabled}
+            end
+    end.
+
+%% Removal the domain name - This function must be idempotent.
+%% domain name must be removed from the core MIM component (if required)
+%% and from the DB. this action must be distributed across
+%% all the nodes in the cluster.
+%% Returns ok, if domain not found.
+%% Domain should be nameprepped using `jid:nameprep'.
+-spec remove_domain(domain()) ->
+    ok | {error, locked} | {error, {db_error, term()}}
+    | {error, service_disabled}.
+remove_domain(Domain) ->
+    case mongoose_domain_core:is_locked(Domain) of
+        true ->
+            {error, locked};
+        false ->
+            case service_domain_db:enabled() of
+                true ->
+                    check_db(mongoose_domain_sql:remove_domain(Domain));
+                false ->
+                    {error, service_disabled}
+            end
+    end.
+
+%% Disabling/Enabling domain name - This function must be idempotent.
+%% the status of the existing domain must be changed.
+%% If domain name is enabled, then it must be added in the core MIM component.
+%% On disabling domain name must be removed from the core MIM component.
+%% Change of the status must be distributed across all the nodes in the cluster.
+-spec disable_domain(domain()) ->
+    ok | {error, not_found} | {error, locked} | {error, duplicate}
+    | {error, service_disabled}.
+disable_domain(Domain) ->
+    case mongoose_domain_core:is_locked(Domain) of
+        true ->
+            {error, locked};
+        false ->
+            case service_domain_db:enabled() of
+                true ->
+                    check_db(mongoose_domain_sql:disable_domain(Domain));
+                false ->
+                    {error, service_disabled}
+            end
+    end.
+
+-spec enable_domain(domain()) ->
+    ok | {error, not_found} | {error, locked} | {error, duplicate}
+    | {error, service_disabled}.
+enable_domain(Domain) ->
+    case mongoose_domain_core:is_locked(Domain) of
+        true ->
+            {error, locked};
+        false ->
+            case service_domain_db:enabled() of
+                true ->
+                    check_db(mongoose_domain_sql:enable_domain(Domain));
+                false ->
+                    {error, service_disabled}
+            end
+    end.
+
+check_db(ok) ->
+    %% Speedup the next check.  %% It's async.
+    service_domain_db:force_check_for_updates(),
+    ok;
+check_db(Result) ->
+    Result.
+
+%% Domain should be nameprepped using `jid:nameprep'
+-spec get_host_type(domain()) ->
+    {ok, host_type()} | {error, not_found}.
+get_host_type(Domain) ->
+    mongoose_domain_core:get_host_type(Domain).
+
+%% Get the list of the host_types provided during initialisation
+%% This has complexity N, where N is the number of online domains.
+-spec get_all_locked() -> [domain()].
+get_all_locked() ->
```

Comment:
not sure if we need this interface at all, but if we do then why not to return host_types as well?

Reply:
it was in the spec...

Reply:
let's return domains' host types as well.